File: test_puzzle.py
print("Puzzle | ZeroBytes | Full Range          | Stripped Range      | Reduction")
print("-------|-----------|---------------------|---------------------|----------")
for puzzle in [71, 72, 73, 74, 76, 80, 90, 100, 120, 135]:
    full_start = 1 << (puzzle - 1)
    full_end = (1 << puzzle) - 1
    full = full_end - full_start

    # Max zero bytes we can strip (must leave at least 1 bit)
    max_strip = (puzzle - 1) // 8
    if max_strip > 0:
        remaining = puzzle - (max_strip * 8)
        strip_start = 1 << (remaining - 1)
        strip_end = 1 << remaining
        stripped = strip_end - strip_start
        reduction = full / stripped
        print(f"  {puzzle:4} | {max_strip:9} | {full:>18} | {stripped:>18} | {reduction:>8.1f}x")
    else:
        print(f"  {puzzle:4} | {0:9} | {full:>18} | {full:>18} | {'1.0':>8}x")

File: test_puzzle.rs
fn puzzle_range(puzzle_num: u32) -> (u128, u128) {
    let start = 1u128 << (puzzle_num - 1);
    let end = (1u128 << puzzle_num) - 1;
    (start, end)
}

fn stripped_range(puzzle_num: u32, zero_bytes: u32) -> (u128, u128) {
    let remaining_bits = puzzle_num - (zero_bytes * 8);
    let start = 1u128 << (remaining_bits - 1);
    let end = 1u128 << remaining_bits;
    (start, end)
}

fn main() {
    println!("Puzzle | ZeroBytes | Full Range         | Stripped Range     | Reduction");
    println!("-------|-----------|--------------------|--------------------|----------");
    for puzzle in [71u32, 72, 73, 74, 76, 80, 90, 100, 120, 135] {
        let zero_bytes = (256 - puzzle) / 8;
        let (fs, fe) = puzzle_range(puzzle);
        let full = fe - fs;
        let remaining = puzzle - (zero_bytes * 8);
        if remaining > 0 {
            let (ss, se) = stripped_range(puzzle, zero_bytes);
            let stripped = se - ss;
            let reduction = full as f64 / stripped as f64;
            println!("  {:4} | {:9} | {:>18} | {:>18} | {:>8.1}x", 
                puzzle, zero_bytes, full, stripped, reduction);
        }
    }
}
